### PR TITLE
PR-Rails-Wallet - Ruby on Rails capstone project  

### PR DESCRIPTION
**Pull request by ivanmvh**
# Rails-Wallet - Ruby on Rails capstone project  

## Description

The Ruby on Rails capstone project ([remember what they are?](https://github.com/microverseinc/curriculum-html-css/blob/main/articles/capstone_intro.md)) is about building a mobile web application where we can manage our budget: we have a list of transactions associated with a category, so that we can see how much money we spent and on what. 

We will create a Ruby on Rails application that allows the user to:
- register and log in, so that the data is private to them.
- introduce new transactions associated with a category.
- see the money spent on each category.

image

#### Design
- We should follow these [design guidelines](https://www.behance.net/gallery/19759151/Snapscan-iOs-design-and-branding?tracking_source=), including:
  - Colors.
  - Typography: font face, size and weight.
  - Lawet: composition and space between elements.

The [Creative Commons license of the design](https://creativecommons.org/licenses/by-nc/4.0/) requires that we give appropriate credit to the author. Therefore, we must do it in the README of our project.

#### Interactions
- Splash screen
  - A simple page with the name of our app (yes, we need to choose one), and links to the sign up and log in pages.

- Sign up and log in pages
  - The user should be able to register in the app with full name, email and password (all mandatory).
  - The user can log into the app using email and password.
  - If the user is not logged in, they can't access pages that require the user to be logged in (all the pages described below).

- Home page (categories page)
  - When the user logs in, they are presented with the categories page.
  - For each category, the user can see their name, icon and the total amount of all the transactions that belongs to that category.
  - When the user clicks (or taps) on a category item, the application navigates to the transactions page for that category.
  - There is a button "add a new category" at the bottom that brings the user to the page to create a new category.

- Transactions page
  - For a given category, the list of transactions is presented, ordered by the most recent.
  - At the top of the page the user could see the total amount for the category (sum of all of the amounts of the transactions in that category).
  - There is a button "add a new transaction" at the bottom that brings the user to the page to create a new transaction.
  - When the user clicks on the "Back" button (<), the user navigates to the home page.

- "Add a new category" page
  - The user fills out a form to create a new category, indicating their name and icon (both mandatory).
  - The user clicks (or taps) the "Save" button to create the new category, and is taken to the home page on success.
  - When the user clicks on the "Back" button (<), the user navigates to the home page.

- "Add a new transaction" page
  - The user fills out a form to create a new transaction with:
    - name (mandatory)
    - amount (mandatory)
    - categories (mandatory at least one)
  - The user click (or taps) the "Save" button to create the new transaction, and is taken to the transactions page for that category.
  - When the user clicks on the "Back" button (<), the user navigates to the transactions page for that category.

#### Testing requirements
- Create unit and integration tests for all the most important components of our RoR application.

#### Technical requirements

- We should use Postgres as our database.
- We should use devise for authentication.
- We should validate all user input to make sure that anyone with bad intentions cannot compromise our app.
- We can use a view template engine of our choice (.erb, .slim, .haml).
- The project should be deployed and accessible online.
- our database schema should reflect the following structure:

## Challenge breakdown

Activity list:

- Set up the repository and tools.
- Create our models according to the ERD diagram. Don't forget about validations and tests.
- Implement the authentication and authorization. 

- Create the page to add a category.
- Create the home page.
- Create the page to add a transaction.
- Create the transactions page.

- Make sure our app is tested adequately.
- Deploy the project, and test for final details.
  [Deploy link](https://rails-wallet.onrender.com)
- Record a video for our project.
  [Video link]()
- Create a good README and PR description.

### General requirements (Don't forget)

- Make sure that there are no linter errors.
- Make sure that we used correct gitflow.
- Make sure that we documented our work in a professional way.
- Follow our list of best practices for Ruby.

@ivanmvh

### CODE REVIEW

File: config/initializers/devise.rb
Comment (mukhammadrizooff) [inline config/initializers/devise.rb:1]: - [ ] [REQUIRED] Good job so far! I appreciate what you're doing, but please `style` it like the design featured on [Behance](https://www.behance.net/gallery/19759151/Snapscan-iOs-design-and-branding?tracking_source=) by `Microverse` 🙏🏼 
* Colors
* Typography: font face, size, and weight.
* Layout: composition and space between elements.

### Your Design
image

### Behance Design
image

File: app/assets/stylesheets/application.css
Comment (mukhammadrizooff) [inline app/assets/stylesheets/application.css:1]: - [ ] [REQUIRED] Good job so far! I appreciate what you're doing, but please `style` it like the design featured on [Behance](https://www.behance.net/gallery/19759151/Snapscan-iOs-design-and-branding?tracking_source=) by `Microverse`🙏🏼
* Colors
* Typography: font face, size, and weight.
* Layout: composition and space between elements.

### Home Design
image


### Details Design
image

File: MIT.md
Comment (mukhammadrizooff) [inline MIT.md:1]: - [ ] [REQUIRED] It would be proffesional if you create your `MIT` license because your `MIT` license link on `README` is not working 🙏🏼 

image

File: spec/devise/sessions_spec.rb
Comment (mukhammadrizooff) [inline spec/devise/sessions_spec.rb:1]: - [ ] [REQUIRED] Good job so far! Please fix the `failure` of this spec file 🙏🏼 

image

**Review by mukhammadrizooff — CHANGES_REQUESTED** [review]
Hi @ivanmvh, 

Good job so far!
There are some issues that you still need to work on to go to the next project but you are almost there!



## Highlights
- No linter errors ✔️ 
- Good implementations ✔️ 
- Correct GitFlow ✔️ 
- Descriptive `commit` history ✔️ 

## Required Changes ♻️
_Check the comments under the review._

## Optional suggestions
- [ ] [OPTIONAL] It would be nice if you replace the link of your presentation video on `PR` summary with your `README`'s one 👍🏼 

Cheers, and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

File: spec/devise/sessions_spec.rb
Comment (ivanmvh) [reply spec/devise/sessions_spec.rb:1]: I dont know why the first time I run some procedures If I run anbother time it runs ok. It has somethin  to do with the windows platform. They say something about temp file writing but as I said this occurs the first time. Run the rspec anothetr time and it run without problem. Thanks for your attention.

File: MIT.md
Comment (BenMukebo) [reply MIT.md:1]: * [ ] [REQUIRED] Dear @ivanmvh, Please remember we are not expected to have any broken links in our readme file, When I click the link to the `MIT.md` file that you added, I get this 404 `error`. t. Kindly as requested by the previous reviewer above 🙃 fix this 👍

File: app/controllers/groups_controller.rb [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+  def create
+    @group = current_user.groups.build(group_params)
+
+    respond_to do |format|
+      if @group.save
+        format.html { redirect_to user_group_path(current_user, @group), notice: 'Group was successfully created.' }
+        format.json { render :show, status: :created, location: @group }
+      else
+        format.html { render :new }
+        format.json { render json: @group.errors, status: :unprocessable_entity }
```
Comment (BenMukebo) [inline app/controllers/groups_controller.rb]: ###  "Add a new category or Group" page

- [OPTIONAL] According to the project requirements, When the user clicks (or taps) the "Save" button to create the new category, it must take the user to the home page on success and not the `group/:id` page. Kindly correct this as well. ✍🏻

File: app/views/groups/index.html.erb [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+              <% else %>
+                <%= image_tag "default_product_image.webp", class: "group-icon" %>
+              <% end %>
+              <h3> <%= group.name %> </h3>
+              <p id="total-amount">$<%= group.entities.sum(:amount) %></p>
+          </div>
+        <% end %>
+    <% end %>
+    <%= link_to "Add New Category", new_user_group_path(current_user), class: "add-btn green-btn" %>
+  </section>
```
Comment (BenMukebo) [inline app/views/groups/index.html.erb]: - [x] [REQUIRED] As requested by the previous reviewer,  Kindly you need to improve your design, Check these [design guidelines](https://www.behance.net/gallery/19759151/Snapscan-iOs-design-and-branding?tracking_source=), including:
* Colors
* Typography: font face, size, and weight.
* Layout: composition and space between elements.
- [x] [REQUIRED] Kindly follow the provided design (width, font, size, height, background, color...)
1. The `ADD NEW TRANSACTION` button must always be at the bottom of your page. ⬇️  
2. Add this white border around the ADD NEW CATEGORY button text. 👍🏻
3. The Add Transaction button has a similar problem. Kindly fix it as well. 💯  👍🏻

- Original
image

- Your Button
image


image

File: app/models/user.rb [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+  def set_default_role
+    role || :user
+  end
+
+  has_one_attached :avatar
+  has_many :entities
+  has_many :groups
+
+  validates :name, presence: true
+  validates :email, presence: true
```
Comment (BenMukebo) [inline app/models/user.rb:19]: - [OPTIONAL] Great job adding user name validation 👏🏻 Let's now validate all user input (password, ...) to make sure that anyone with bad intentions cannot compromise their app.

File: spec/factories/entities.rb [hunk 4]
```diff
@@ -0,0 +1,11 @@
+require 'faker'
+
+FactoryBot.define do
+  factory :entity do
+    name { Faker::Company.name }
+    amount { Faker::Number.decimal(l_digits: 2) }
```
Comment (BenMukebo) [inline spec/factories/entities.rb:6]: - [OPTIONAL] Excellent addition of unit tests for most of the most important components of the RoR application. 👍🏻 
 Now, let's make sure to also add integration tests with capybara for all the most important components of the RoR application. (movements) ✍🏻

image

File: app/views/entities/new.html copy.erb [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+      p "@entity"
+      p @entity
+      <% if @entity.errors.any? %>
+        <div id="error_explanation">
+          <h2><%= pluralize(@entity.errors.count, "error") %> prohibited this transacction from being saved:</h2>
+          <ul>
+            <% @entity.errors.full_messages.each do |message| %>
+              <li class="alert-danger"><%= message %></li>
+            <% end %>
+          </ul>
```
Comment (BenMukebo) [inline app/views/entities/new.html copy.erb:13]: ### Transactions page
- [ ] [OPTIONAL] Kindly make add the "Back" button (<) on the Transactions page, so that When the user clicks on the "Back" button the user navigates to the home page.

image



### Add a new transaction page

- [OPTIONAL]  According to the project requirements, When the user clicks on the "Back" button (<), the user navigates to the transactions page for that category and not the Home Page. Kindly correct this as well. ✍🏻 
- [OPTIONAL] It's one of the project requirements that The user fills out a form to create a new transaction with these mandatory parameters:
* name (mandatory)
* amount (mandatory)
* categories (mandatory at least one). 

image


- [OPTIONAL] Kindly add validation to the categories input select field to prevent any user to submit the form without selecting any category👍🏻

File: db/migrate/20230503183426_create_entities.rb [hunk 6]
```diff
@@ -0,0 +1,11 @@
+class CreateEntities < ActiveRecord::Migration[7.0]
+  def change
+    create_table :entities do |t|
+      t.string :name
+      t.integer :author_id
+      t.decimal :amount
+
+      t.timestamps
```
Comment (BenMukebo) [inline db/migrate/20230503183426_create_entities.rb:8]: - [OPTIONAL] The database schema should reflect the following structure:

image

- [OPTIONAL] According to the Entity relational database Diagram, there is a many to many relationship between Group and Entity so, You should have created a join table to link the two tables. Kindly address this 💯 😉 
- Note: In Rails and PostgreSQL, creating a join table is necessary for modeling many-to-many relationships because it allows us to establish a connection between two tables without duplicating data.

File: config/routes.rb [hunk 7]
```diff
@@ -0,0 +1,18 @@
...
+  get 'entities/new'
+  get 'entities/destroy'
+  get 'entities/edit'
+  get 'entities/update'
+  get 'entities/index'
+  devise_for :users, sign_out_via: [:get, :post], controllers: { registrations: "users/registrations" }
+  resources :users, only: [:index, :show, :destroy] do
+    resources :groups do
+      resources :entities
+    end
```
Comment (BenMukebo) [inline config/routes.rb:12]: - [OPTIONAL] It would be nice to use only and except options to add only necessary routes in case maybe there are some routes that are not used. 👍🏻

**Review by BenMukebo — CHANGES_REQUESTED** [review]
## Hi @ivanmvh, 

You did a great job so far 👏🏼 👏🏼
There are some issues that you still need to work on to go to the next project but you are almost there!
<FEEL FREE TO CHANGE THIS MESSAGE TO YOUR PERSONAL WORDS>


### To Highlight 🎉
- PR has a good title and summary. ✔️
- Correctly use Devise for authentication. ✔️ 
- Your app has been well deployed. ✔️

## STATUS: CHANGES REQUESTED  ♻️ 
_Please Check the comments under the review and make all required changes_
## Optional suggestions

_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing them will really improve your application._

Cheers :clinking_glasses:  and Happy coding!!! :dancers:💻  

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please ping me @BenMukebo when you comment so I can receive the notification.


------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

File: app/views/entities/new.html copy.erb (hunk 5 above)
Comment (KayLemba) [inline app/views/entities/new.html copy.erb:13]: [OPTIONAL]As the previous code reviewer request please kindly implement this change I know it an optional change but it should not be overlooked because it will help make your code better.

- Kindly make add the "Back" button (<) on the Transactions page, so that When the user clicks on the "Back" button the user navigates to the home page.

236657956-bcca9520-e88a-4759-ad5d-aad2370c3cae

Your design should look like this:

Screenshot 2023-05-07 at 7 25 04 AM

it is important to pay attention to detail follow design requirements

**Review by KayLemba — APPROVED** [review]
[[# STATUS: APPROVED :green_circle: :clap: 

Hello @ivanmvh      , 
Your project is complete!  There is nothing else to say other than... it's time to merge it :shipit: 
Congratulations! 🎉  Keep up the good work 🥂 

HIGHLIGHTS:

- [x]  Your repo is professional  😎 
- [x] Descriptive P.R 💯 
- [x] linters are passing
- [x] Professional README 💯
- [x] App running as intended  🥇  
- [x] Tests are passing 

 










 








Now you can merge this branch and move on :1st_place_medal: 
Happy Coding cheers 🥇 💯 



well-1

Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better.

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._]]

File: app/views/entities/new.html.erb [hunk 8]
```diff
@@ -0,0 +1,49 @@
...
+        <div class="field">
+          <%= check_box_tag 'Category', 'group.id', true, disabled: false %>
+          <%= f.label @group.name %>
+        </div>
+
+        <%= f.hidden_field :author_id, value: current_user.id %>
+      </div>
+      <div class="green-rectangle">
+        <div class="white-outline">
+            <%= f.submit 'SAVE TRANSACTION' , class: 'green-btn' %>
```
Comment (GzimAsani) [inline app/views/entities/new.html.erb]: - [x] Here a ```back``` button should be present which will navigate the user to the transactions page for that category
image

File: app/views/entities/new.html copy.erb [hunk 9]
```diff
@@ -0,0 +1,43 @@
...
+      </div>
+
+      <%= f.hidden_field :author_id, value: current_user.id %>
+
+      <div class="field">
+        <% @groups.each do |group| %>
+          <p>
+              <%= f.check_box :group_id, { multiple: true, checked: false, hidden: false }, group.id, nil %>
+              <%= f.label group.name %>
+          </p>
```
Comment (GzimAsani) [inline app/views/entities/new.html copy.erb:34]: - [x] Here I am seeing only one category from where I clicked to make the transaction, according to the requirements I should see all the present categories and I can select multiple of them 
image
image

File: app/models/entity.rb [hunk 10]
```diff
@@ -0,0 +1,8 @@
+class Entity < ApplicationRecord
+  belongs_to :group
```
Comment (GzimAsani) [inline app/models/entity.rb]: -  [ ] Entities should also  have_many groups, this is a project requirement, kindly check the provided ERD diagram. Their relationship should be many to many 
image

File: spec/models/entity_spec.rb [hunk 11]
```diff
@@ -0,0 +1,13 @@
+require 'rails_helper'
+
+RSpec.describe Entity, type: :model do
+  it 'has a valid factory' do
+    expect(build(:entity)).to be_invalid
```
Comment (GzimAsani) [inline spec/models/entity_spec.rb:5]: - [ ] I think here you should test to be_valid instead

File: spec/factories/entities.rb (hunk 4 above)
Comment (GzimAsani) [inline spec/factories/entities.rb:6]: - [ ] Please include integration tests currently you have none. Also be reminded: 
image

File: app/views/entities/new.html copy.erb (hunk 9 above)
Comment (GzimAsani) [reply app/views/entities/new.html copy.erb:34]: - Kindly add validation to the categories input select field to prevent any user to submit the form without selecting any category

File: app/views/entities/index.html.erb [hunk 12]
```diff
@@ -0,0 +1,41 @@
...
+<section class="container">
+    <% total_amount = 0 %>
+    <% @entities.each do |entity| %>
+          <% total_amount += entity.amount %>
+    <% end %>
+    <table>
+      <thead>
+      <tr>
+          <th> </th>
+          <th>Total Amount</th>
```
Comment (GzimAsani) [inline app/views/entities/index.html.erb:19]: - [x] Categories show page should follow this design 
image

**Review by GzimAsani — CHANGES_REQUESTED** [review]
### Hello, @ivanmvh                 :wave:   
 Good job so far,  done There are some issues that you still need to work on to go to the next project but you are almost there!
 
 ## Highlights
 * Good job so far with your styling, looks neat

 ## Required Changes 

- [ ] Check the comments below :arrow_double_down:  

 ## Optional suggestions
 _Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._



Cheers, and Happy coding! clapclapclap
 
 Feel free to leave any questions or comments in the PR thread if something is not 100% clear. Please, remember to tag me in your question so I can receive the notification.
 
 **_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**
 
 _As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

File: app/views/entities/new.html.erb (hunk 8 above)
Comment (ivanmvh) [reply app/views/entities/new.html.erb]: the back button was added.  Thanks

File: app/views/entities/new.html copy.erb (hunk 9 above)
Comment (ivanmvh) [reply app/views/entities/new.html copy.erb:34]: If not selected, it assumes the categories of the displayed page.

File: app/views/entities/index.html.erb (hunk 12 above)
Comment (ivanmvh) [reply app/views/entities/index.html.erb:19]: There is not a show page for a category. There is a list of transactions (no one but the most recent transactions up) . I will adapt the list to this design. 
Here the requirement:
**Home page (categories page**)
When the user clicks (or taps) on a category item, the application navigates to the transactions page for that category. 

**Transactions page**
    For a given category, the list of transactions is presented, ordered by the most recent.
    At the top of the page the user could see the total amount for the category (sum of all of the amounts of the transactions in that category).
    There is a button "add a new transaction" at the bottom that brings the user to the page to create a new transaction.
    When the user clicks on the "Back" button (<), the user navigates to the home page.
